File: Tree/Leetcode-110-BalancedBinaryTree/Leetcode-110-BalancedBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if( root == nullptr ) return true;
        int diff = getHeight(root->left) - getHeight(root->right);
        if( diff > 1 || diff < -1 ) return false;
        return isBalanced(root->left) && isBalanced(root->right);
    }
    int getHeight(TreeNode* sub) {
        if( sub == NULL ) return 0;
        return max(getHeight(sub->left), getHeight(sub->right))+1;
    }
};

File: 2월4주차/leetcode-110-Balanced Binary Tree.py
# Definition for a binary tree node
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution: # 내 풀이
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        unbalanceFlag = 0
        def dfs(node): # 자신을 root로 했을 때 tree의 깊이를 return함. None이면 0, 자기만 있으면 1, ..
            nonlocal unbalanceFlag
            if not node or unbalanceFlag:
                return 0

            ld, rd = dfs(node.left), dfs(node.right)
            if abs(ld-rd) > 1:
                unbalanceFlag = 1
                return 0
            return max(ld, rd) + 1

        dfs(root)
        if unbalanceFlag:
            return False
        else:
            return True
        
class Solution: # 솔루션 : unbalancedFlag 대신 dfs함수가 -1을 return하면 불균형한 것으로 약속
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(node): # 자신을 root로 했을 때 tree의 깊이를 return함. None이면 0, 자기만 있으면 1, ..
            if not node:
                return 0

            ld, rd = dfs(node.left), dfs(node.right)
            if ld == -1 or rd == -1 or abs(ld-rd) > 1:
                return -1

            return max(ld, rd) + 1

        if dfs(root) == -1:
            return False
        else:
            return True